onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc

```diff
@@ -2436,6 +2436,70 @@ Status FlattenOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder, cons
 
 #pragma endregion
 
+#pragma region op_gather
+
+class GatherOpBuilder : public BaseOpBuilder {
+ public:
+  void AddInitializersToSkip(ModelBuilder& model_builder, const NodeUnit& node_unit) const override;
+
+ private:
+  Status AddToModelBuilderImpl(ModelBuilder& model_builder, const NodeUnit& node_unit) const override;
+};
+
+void GatherOpBuilder::AddInitializersToSkip(ModelBuilder& model_builder, const NodeUnit& node_unit) const {
+  // Skip the second input `indices` for Gather
+  const auto& inputs = node_unit.Inputs();
+  model_builder.AddInitializerToSkip(inputs[1].node_arg.Name());  // indices
+}
+
+Status GatherOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder, const NodeUnit& node_unit) const {
+  auto& shaper(model_builder.GetShaper());
+  const auto& operand_indices(model_builder.GetOperandIndices());
+  const auto& operand_types(model_builder.GetOperandTypes());
+  const auto& initializers(model_builder.GetInitializerTensors());
+  const auto& input1 = node_unit.Inputs()[0].node_arg.Name();
+  const auto& input2 = node_unit.Inputs()[1].node_arg.Name();  // "indices"
+  const auto& output = node_unit.Outputs()[0].node_arg.Name();
+
+  NodeAttrHelper helper(node_unit);
+  int32_t rank = static_cast<int32_t>(shaper[input1].size());
+  int32_t axis = static_cast<int32_t>(HandleNegativeAxis(helper.Get("axis", 0), rank));
+
+  std::vector<uint32_t> input_indices;
+  input_indices.push_back(operand_indices.at(input1));
+  ADD_SCALAR_OPERAND(model_builder, input_indices, axis);
+
+  // Add indices operand into nnapi
+  const auto& indices_tensor = *initializers.at(node_unit.Inputs()[1].node_arg.Name());
+  std::vector<uint8_t> unpacked_tensor;
+  ORT_RETURN_IF_ERROR(onnxruntime::utils::UnpackInitializerData(indices_tensor, unpacked_tensor));
+  const int64_t* raw_indices_data = reinterpret_cast<const int64_t*>(unpacked_tensor.data());
+
+  const auto indices_shape = indices_tensor.dims();
+  uint32_t size = 1;
+  Shape indices_dimen;
+  indices_dimen.reserve(indices_tensor.dims_size());
+  for (auto i = 0; i < indices_tensor.dims_size(); i++) {
+    size *= indices_shape[i];
+    indices_dimen.push_back(static_cast<uint32_t>(indices_shape[i]));
+  }
+  std::vector<int32_t> indices(size);
+  for (uint32_t i = 0; i < size; i++) {
+    indices[i] = SafeInt<int32_t>(raw_indices_data[i]);
```

Comment:
I'm not so sure about using reinterpret_cast for raw_indices_data. It's kind of tricky to use without undefined behavior.
An alternative:
```suggestion
    int64_t index_i64;
    memcpy(&index_i64, unpacked_tensor.data() + i * sizeof(int64_t), sizeof(int64_t));
    indices[i] = SafeInt<int32_t>(index_i64);
```

https://gist.github.com/shafik/848ae25ee209f698763cffee272a58f8#type-punning-arrays

Reply:
Thanks for asking this. It is necessary to consider both int32/int64 type indices here. The unit tests do not cover a case with float input data + int32 indices,  overlooked it before.

Do you have a suggested good way to accommodate both cases here?

Reply:
Added additional unit test to separately to cover int32_t indices case with float input data for NNAPI.

Reply:
Can we reference the gist in a comment here to explain the use of memcpy?